mlflow/mlflow
# Fix artifact URI constructions in FileStore and SqlAlchemyStore

## What changes are proposed in this pull request?

Currently, the FileStore and SqlAlchemyStore construct experiment and run artifact root URIs by appending the experiment ID or run artifact path to the store's artifact root URI via the `posixpath.join` function. Unfortunately, this behavior is not aware of URI structures such as queries and fragments. As a result, for an artifact root of `db://root:password@localhost:port/dbname?driver=odbc`, experiment URIs take the form `db://root:password@localhost:port/dbname?driver=odbc/{experiment_id}`, and run artifact URIs take the form `db://root:password@localhost:port/dbname?driver=odbc/{experiment_id}/{run_id}/artifacts`. We see that the experiment and run information is being appended to the *query* component of the URI, rather than the path.

This PR fixes the behavior to correctly append experiment and run details to the *path* component of artifact root URIs. For the case above, the stores will now construct experiment and run artifact URIs as follows:
```
db://root:password@localhost:port/dbname/{experiment_id}?driver=odbc
db://root:password@localhost:port/dbname/{experiment_id}/{run_id}/artifacts?driver=odbc
```

## How is this patch tested?

Thorough unit tests, extensive manual testing

**Manual test plan**
This PR was manually tested by verifying the correct behavior of the MLflow artifact and model . tracking APIs against artifact stores with the following URIs:
- The current working directory (local filesystem)
- A specific local filesystem directory ("/tmp/directory@")
- An Amazon S3 bucket location of the form ("s3://bucket_name/artifact/root/path")
- A Microsoft SQL Server database running in a Docker container with the following URI: `mssql+pyodbc://root:password@localhost:1433/dbname?driver=ODBC+Driver+17+for+SQL+Server`, after merging these changes with a copy of the pending DB artifact storage plugin branch (https://github.com/dbczumar/mlflow/tree/dbartifacts_uri_update).

For each case, I ran the following code (replacing the artifact location with the specified URI):
```
import mlflow
import mlflow.pyfunc
from mlflow.tracking.artifact_utils import _download_artifact_from_uri
from mlflow.tracking.client import MlflowClient
from mlflow.pyfunc import PythonModel

client = MlflowClient()

exp = client.create_experiment("s3exp", artifact_location="s3://artifact-testing-for-mlflow-bucket/subpath/root")


mlflow.set_experiment("s3exp")

with mlflow.start_run():

    class Mod(PythonModel):
        
        def predict(self, ctx, inp):
            return 7

    mlflow.pyfunc.log_model(
        python_model=Mod(),
        artifact_path="mymodel")

    model_uri = "runs:/{run_id}/mymodel".format(run_id=mlflow.active_run().info.run_id)

print(mlflow.pyfunc.load_model(model_uri))

with mlflow.start_run():
    local_path = "/tmp/myartifact.txt"
    with open(local_path, "w") as f:
        f.write("MY FILE")

    mlflow.log_artifact(local_path, "artifact/sublocation")
    artifact_uri = "runs:/{run_id}/artifact/sublocation/myartifact.txt".format(run_id=mlflow.active_run().info.run_id)
    downloaded_location = _download_artifact_from_uri(artifact_uri)
    with open(downloaded_location, "r") as f:
        print(f.read())
```

I then verified that artifacts were written to the appropriate storage locations (e.g., S3, local filesystem, SQL server) and that the script produced correct output of the form:
```
<mlflow.pyfunc.model._PythonModelPyfuncWrapper object at 0x1100051d0>
MY FILE
```

## Release Notes

### Is this a user-facing change?


### What component(s) does this PR affect?

- [X] Artifacts
- [ ] Scoring
- [ ] Serving
- [ ] R

### How should the PR be classified in the release notes? Choose one:
